feat: Horizontal Pod Autoscaler

dvanmali opened the pull request:
Add an optional chart configuration for Horizontal Pod Autoscaler for the Surreal Deployment.

Changes chart from 0.3.6 > 0.3.7

mumoshu commented on charts/surrealdb/templates/hpa.yaml:
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.horizontalPodAutoscaler}}
```
There could be various conventions here, but I believe the most popular is to have an explicit helm chart value to toggle the HPA:

```suggestion
{{- if .Values.autoscaling.enabled }}
```

dvanmali replied:
It may not be necessary to add an autoscaling enabled flag. I can make the changes such that only when the horizontal pod autoscaler variables are defined, then the autoscaling feature is enabled, otherwise it uses replicaCount

mumoshu replied:
Thanks for your response! Yeah, it may not be necessary. I just believe using the enabled flag pattern is better.

By using the existence of a non-empty value to enable the feature, you end up with multiple`.Values.foo | default ...` in the chart templates, like `minReplicas: {{ .minReplicas | default 1 }}` we can see in the diff. That's repetition because we write down the same default values in two other places, the values.yaml comments and README.

With the `.enabled` pattern, you can declare and document all the defaults in the single-place- chart's default values.yaml file- and we can generate the README.md from the values.yml using `helm-docs` in the future, leaving us no need to touch the defaults in the templates and manually update the docs.

dvanmali replied:
Okay added `.Values.horizontalPodAutoscaler.enabled` pattern to be compatable with helm-docs

mumoshu reviewed (changes requested):
He @dvanmali! We appreciate your contribution!

This looks great, but I believe we need to address a thing correctly to move forward.

That is, this doesn't work, until we omit the deployment's desired replicas when HPA is enabled when HPA is active. Please see https://github.com/surrealdb/helm-charts/blob/991915b451939ba97b0884c86bb80c3d60feade0/charts/surrealdb/templates/deployment.yaml#L8

Note that it's not straightforward when you deploy the chart without HPA and then enabled it in a second deployment- It's not handled by Helm well so you need some manual steps to make it work. Please also see https://github.com/helm/helm/issues/12650 for more details.

It would be great if you could:

- Fix the chart so that the deployment replicas is omitted when HPA is active
- Have some guidance in the chart README about how to migrate from non-HPA deployment to HPA-enabled

Thank you in advance for your support!

dvanmali commented:
Good catch @mumoshu and thanks for reviewing this PR :)

Fixed the HPA file and added docs to state that HPA is an optional feature which can be added/removed at anytime using `helm update`.

dvanmali commented:
@mumoshu Fixed with an enabled flag now.

I also ran `helm-docs` locally on my computer and some of the docs did not match. Note that the [README.md](charts/surrealdb/README.md) is fixed to match.

mumoshu reviewed (approved):
I appreciate your effort, @dvanmali!
This LGTM!

I manually verified that it supports turning HPA on/off. I've also added a quick E2E test to prevent it from breaking in the future. I hope you like it.

Having two or more pods works only when you have a distributed backend like TiKV and FoundationDB. We should document this in a separate pull request.

Thank you for your contribution, @dvanmali!

dvanmali commented:
Much thanks for writing the e2e tests and the merge 🥇  Looks great! 😄
